src/dbnode/storage/shard.go

```diff
@@ -2854,6 +2871,38 @@ READER:
 	return processedTileCount.Load(), nil
 }
 
+func encodeDownsampledValues(
+	downsampledValues []tile.DownsampledValue,
+	frame tile.SeriesBlockFrame,
+	ant ts.Annotation,
+	encoder encoding.Encoder,
+) error {
+	for _, downsampled := range downsampledValues {
+		timestamp := frame.Timestamps()[downsampled.FrameIndex]
+		dp := ts.Datapoint{
+			Timestamp:      timestamp,
+			TimestampNanos: xtime.ToUnixNano(timestamp),
+			Value:          downsampled.Value,
+		}
+
+		// TODO: what happens if unit has changed mid-tile?
+		// Write early and then do the remaining values separately?
+		unit, err := frame.Units().Value(downsampled.FrameIndex)
+		if err != nil {
+			return err
+		}
```

Comment:
nit: why are we doing units per index, but annotations taken from the first value in the frame? Should do the same thing for both here; imo use the first seen unit and annotation for the entire block

Reply:
👍